### DISCUSSION TITLE
Save a pdf with form fields disabled.

### DISCUSSION THREAD

**Discussion by adilsitos** [pymupdf/PyMuPDF#2323] [Looking for help] [answered]
I am using pymupdf to fill a pdf with form fields (widget), and it helped me a lot! But when I save the pdf it is still possible to change these form fields, is there a way to save this PDF with the widgets disabled?

**Comment by JorjMcKie** [comment]
You mean that any later open of the PDF won't allow to change the PDF (and thus the fields)?
If yes, you could provide master and user passwords on save and also set the permissions such that opening with the user password prohibits changes.

**Reply by JorjMcKie** [reply]
Look up permissions flags bit [here](https://pymupdf.readthedocs.io/en/latest/vars.html#document-permissions). In your case you could use a value with `fitz.PDF_PERM_ANNOTATE` set off.
In addition, there also is an annotation flag bit `fitz.PDF_ANNOT_IS_LOCKED`. This must be set for the field's underlying annotation.

**Reply by adilsitos** [reply]
Thanks for your reply. I tried to save using the following command: 
`doc.save("filename.pdf", permission=fitz.PDF_PERM_ANNOTATE)` but it is still possible to change the fields in this new PDF. Did I used it in the wrong way? Regarding the annotation flag you mentioned, I didn't really understand, it should be used in the widget itself? I tried to use in a widget like this: 
`widget.field_flag = fitz.PDF_ANNOT_IS_LOCKED`

**Reply by adilsitos** [reply]
I figure out a way. I used the flag `fitz.PDF_FIELD_IS_READ_ONLY` on the changed widgets. However, I think I found a bug. My code had this syntaxe: `widget.field_flag` but `field_flag` does not exist. Surprisingly, it didn't return an error.

**Reply by JorjMcKie** [reply, marked as answer]
The correct name is `Widget.field_flags` ("s" at the end). If you want to set a bit, please remember to use Python's "|" operator, **do not use** `widget.field_flags = ...`, because this will have all sorts of side effects, among them changing the field type!

**Reply by adilsitos** [reply]
Thank you. I will change my code to use the correct operator!

**Reply by adilsitos** [reply]
So I changed the operator, but in the checkbox widgets the read only flag (`fitz.PDF_FIELD_IS_READ_ONLY`) does not work, is there a limitation for this type of widget? Text widget works great!

**Reply by JorjMcKie** [reply]
> So I changed the operator, but in the checkbox widgets the read only flag (`fitz.PDF_FIELD_IS_READ_ONLY`) does not work, is there a limitation for this type of widget? Text widget works great!

There is a bug in handling the field flags which affects button fields. This fix is awaiting to be published in the coming version, due in 2 weeks or so.
